rflx/generator/allocator.py

```diff
@@ -216,29 +239,48 @@ def _create_init_proc(self) -> UnitPart:
     def _needs_allocation(type_: rty.Type) -> bool:
         return isinstance(type_, (rty.Message, rty.Sequence))
 
-    def _allocate_slots(self) -> None:
-        """Create memory slots for each construct in the session that requires memory."""
-        count: int = 0
+    def _create_slot_with_size(self, size: int) -> int:
+        slot_num = len(self._size_map) + 1
+        self._size_map[slot_num] = size
+        return slot_num
```

Comment:
Note that the allocation map is used for the code generation (see `get_slot_ptr` and `free_buffer`). We can't just remove it.
As for the the `create_slot_with_size` function, I think this is a standard design to get a fresh id for something. I did some slight refactoring in 36e402566cbeb3d99e3bb7f268abd41e6a87b8f9, but I just realized I left the `_global_slots` variable in, which can be removed now. Will follow up with another commit.

Reply:
The design is still unchanged, `self._size_map` (now called `self._slot_sizes`) and `self._global_slots` are still used to pass data between initialization functions. You are right about `_allocation_map` (now `_allocation_slots`).

Reply:
Yes, I do think my design is a little bit more flexible. The data for sizes, and which slot is a global slot, could become useful later for other things than the generation of the Allocator package.
Given that `_allocation_slots` is needed anyway, I believe there is no extra burden in terms of understanding the flow. The two `_allocate*slots` functions set up the entire state. The `create*` functions use it.

Reply:
I don't think your design has any more flexibility. Instance variables are essentially global variables scoped to an instance. You don't get more flexibility by using global variables instead of parameters and return values. If we ever see the need for using the data outside the Allocator instance, the return value of an initialization function could still be stored as an attribute. The data flow of this class might be simple and obvious to you now, but that doesn't mean that it stays the same in the future. Using some form of global variables is justifiably a bad thing. We have seen similar things in the past. Design flaws that seemed negligible. But more and more design flaws accumulated, and in the end we had code that was barely maintainable.

Reply:
I just don't think that your design is objectively better (which is what you are suggesting). The two designs are just different. At some point the author of the pull request needs to have some liberty on how to implement things. Especially as here everything is local to this rather small class.
If you prefer, we can put the encapsulate the three related variables `allocation_slots`, `slot_sizes` and `global_slots` in a separate class (dataclass?). That would clarify the data flow I think. It would reduce the number of global variables that are manipulated during initialization to one.

Reply:
I don't think both designs are just different. The currently chosen approach has clear disadvantages when it comes to understanding the code base. I had trouble especially in comprehending the side effects of `_allocate_{global|local}_slots` used in the `AllocateGenerator` constructor. I found it very hard to see what the code is doing, especially in places like `_allocate_global_slots` where the result of `_create_slots_with_size` is used as if it was a pure function, but at the same time it has a side-effect on `_size_map`. Being more flexible in the future is a weak argument here: We don't even know whether we will need this flexibility in the future (probably not!) and still we pay the price in terms of decreased comprehensibility. That's a bad deal that we're making here.

Don't misunderstand this as nitpicking. Maintainability of a software is not gained or lost by a single change. It's a series of small degradations that eventually make a software unmaintainable. If we are not strict in avoiding those, we will end up with a code base that where refactorings and extensions are very painful or outright impossible.

Please rework the code to improve it's structure. Ideally, the constructor should use pure functions that e.g. return a list of slot allocations.

Reply:
Thanks for weighing in Alex. I understand this is not nitpicking, that is, you really believe your design is better. I just don't understand the criteria you seem to apply, so it's quite difficult for me to follow them.

In at attempt to rework the design, let me suggest some changes and I need your comments as to which solution you like.

- (A) the current design, for reference
- (B) like (A), but group the three global variables `allocation_slots`, `slot_sizes` and `global_slots` in a separate class `AllocatorData` (naming TBD). `create_slot_with_size` and `map_location_to_slot` would be the "public" interface of that class (public in quotes as `AllocatorGenerator` is the only client). An empty instance of `AllocatorData` is created in the constructor of AllocatorGenerator and then used directly (as a global variable). The `allocate*` functions still have a global side effect, but it's a single global variable (the instance of `AllocatorData`).

- (C) like (B), but the instance of `AllocatorData` is passed to `allocate_global_slots` and `allocate_local_slots` as a parameter, which modify it (via its public interface), and then to the `create*` functions, which read it. The `allocate*` functions still have side effects, but to their parameter, not a global variable.

- (D) the two functions `allocate_*_slots` don't read nor modify any global state, just the session. They each return some data structure which contains the required info, conceptually a list of tuples `(global_or_not, location, size)` (or two lists for global and local allocation points). A separate step assigns slot numbers to those, so we get a list of tuples `(global_or_not, location, size, slot_number)`. We could create one or two new dataclasses to hold this data, but note that both structures are different from the `AllocatorData` above and the already existing `AllocationRequirement`. This last structure is then passed to the `create*` functions, and also used to create the `allocation_map`, possibly in a new private method.

Variant (D) is what I understand you prefer. Or maybe you want to suggest yet another variant?

I personally prefer (B) or (C) (slight preference for B) and I can agree that both are slightly better than the current design (A) as they group the state in a unit that's easier to understand. I dislike (D) despite the fact that is avoids the side effects because of the additional complexity of the intermediate data structures.

Reply:
>  I found it very hard to see what the code is doing, especially in places like _allocate_global_slots where the result of _create_slots_with_size is used as if it was a pure function, but at the same time it has a side-effect on _size_map

I am really surprised you found this part difficult to follow. I believe it is a standard design if you want to create a fresh ID for something. You increment a global counter and return the new value of the counter as the fresh ID. Here the length of the `sizes` map is used as the counter (could use an array, but it would start at 0 ...).

I guess if you don't like this pattern, another variant (E) can be imagined. We could get rid of the slot numbers entirely, and try to derive the name of the slot from the allocation point (currently we use source locations to identify allocation points). The allocation map would not be required as the right slot can be derived from the allocation point. The downside would be awkward variable names in the generated code, that would also change very often even for slight input changes (e.g. adding a comment will change source locations). Personally I dislike (E) for this reason.

Reply:
Final comment: in all the variants I described, I think the flow of data could be simplified by unifying the two `allocate_*_slots` functions; they have been separated in a previous review round.

Reply:
> I just don't understand the criteria you seem to apply, so it's quite difficult for me to follow them.

For me, the main issue with the current design is the unnecessary global side effects of the initialization functions. I'm surprised that you don't see any issue in using globals for passing data between functions.

> In at attempt to rework the design, let me suggest some changes and I need your comments as to which solution you like.

Here are my thoughts to your proposed solutions:

(B) Composing the data and the associated functions into a separate class is an improvement, but I still dislike the global side effects in this variant.
(C) As the global side effects are removed, this would be an acceptable solution for me.
(D) This is conceptually the variant I had in mind. I still think that the data flow is clearest with this variant, even if it is somewhat more cumbersome.
(E) I dislike the negative effects on the generated code.

> in all the variants I described, I think the flow of data could be simplified by unifying the two `allocate_*_slots` functions; they have been separated in a previous review round.

I think it depends on the complexity of those two functions. If they are rather simple, merging them would be fine. But I see no positive impact on the data flow.

Reply:
Thanks, Tobias. Based on your answer I will prepare a commit using solution (C) unless there are other comments.

> I'm surprised that you don't see any issue in using globals for passing data between functions.

Sure I do see issues with that. But it depends on the situation.

Reply:
I tried (C) and found it to be not satisfactory. The `create*` functions need too much info from the `AllocatorData` structure so that the indirection seems artificial and adds quite some noise. (B) would have the same issue.
I still have a slight preference for (A), but given that's not acceptable for you, I implemented (D) in 417914a29fc89919aedc864a67ac772ab949aa0d.